compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -246,24 +279,143 @@ validatePkgId pkgId = do
         fail $ "Invalid package ID dependency in daml.yaml: " <> pkgId
     pure $ LF.PackageId $ T.pack pkgId
 
+-- Package resolution
+---------------------
+
+data FullPkgName = FullPkgName
+    { pkgName :: LF.PackageName
+    , pkgVersion :: LF.PackageVersion
+    } deriving (Eq, Ord, Show)
+
+data LockFile = LockFile
+    { dependencies :: [DependencyInfo]
+    } deriving Generic
+instance Aeson.FromJSON LockFile
+instance Aeson.ToJSON LockFile
+data DependencyInfo = DependencyInfo
+    { name :: String
+    , version :: String
+    , pkgId :: String
+    } deriving Generic
+instance Aeson.FromJSON DependencyInfo
+instance Aeson.ToJSON DependencyInfo
+
+resolvePkgs :: NormalizedFilePath -> Options -> [FullPkgName] -> IO (M.Map FullPkgName LF.PackageId)
+resolvePkgs projRoot opts pkgs
+    | null pkgs = pure M.empty
+    | otherwise = do
+        mbRes <- resolvePkgsWithLockFile lockFp pkgs
```

Comment:
What happens if nothing changed but the lock file is not sufficient. Will `daml build` keep refetching everything on every run and fail?

Reply:
No, it will re-fetch once and write the new lock-file.